### DISCUSSION TITLE
Lithium Plating Submodel if statement

### DISCUSSION THREAD

**Discussion by ch662** [pybamm-team/PyBaMM#2625] [Q&A] [answered]
Hello, I am working on a submodel for lithium plating from the O'Kane 2022 submodel and I wanted to use a Butler-Volmer equation that turns on and off depending on the overpotential like this:

<img width="299" alt="truncated 162651" src="https://user-images.githubusercontent.com/108564982/214973332-3298e632-1c40-49fe-be40-5c842df260b7.png">

But when I try to use an if..then statement in the submodel with eta_stripping I get the error: 

File ~\env\lib\site-packages\pybamm\expression_tree\symbol.py:660, in Symbol.__bool__(self)
    659 def __bool__(self):
--> 660     raise NotImplementedError("Boolean operator not defined for Symbols.")

NotImplementedError: Boolean operator not defined for Symbols.

What can I do to get around this/what am I doing wrong?
Any feedback would be greatly appreciated. Thank you!

**Comment by brosaplanella** [comment, marked as answer]
Instead of the if statement, try multiplying each of the statement with `(eta_stripping >= 0)` and `(eta_stripping < 0)` which should define a heaviside function that does what you want.

**Reply by Hanyu-B** [reply]
Hello, I am Hanyu and I also have a question related to this issue:
 
 Are there any specific reasons why the conditions of lithium plating and SEI growth (eta_stripping should be less than zero ) are not added to the current model? Like issues with the solver? 

Also, the following error occurs when adding the following code to the lithium plating function:
j_stripping=j_stripping * (eta_stripping < 0) + 0 * (eta_stripping >= 0)

But it seems like the bug only occurs at certain cycles where the value of the function changes from zero to one.


psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
The residual routine or the linear setup or solve routine had a recoverable error, but IDACalcIC was unable to recover.
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
The residual routine or the linear setup or solve routine had a recoverable error, but IDACalcIC was unable to recover.
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
The residual routine or the linear setup or solve routine had a recoverable error, but IDACalcIC was unable to recover.
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed
The residual routine or the linear setup or solve routine had a recoverable error, but IDACalcIC was unable to recover.
2023-07-13 11:19:16.073 - [ERROR] callbacks.on_experiment_error(224): Simulation error: Maximum number of decreased steps occurred at t=1704368.2001355456 (final SolverError: '.../casadi/interfaces/sundials/idas_interface.cpp:591: IDACalcIC returned "IDA_NO_RECOVERY". Consult IDAS documentation.'). For a full solution try reducing dt_max (currently, dt_max=600) and/or reducing the size of the time steps or period of the experiment. Set `return_solution_if_failed_early=True` to return the solution object up to the point where failure occured.
2023-07-13 11:19:16.075 - [NOTICE] callbacks.on_experiment_end(220): Finish experiment simulation, took 4 minutes, 32 seconds
Code finished running at: 2023-07-13 11:19:16.207717

**Reply by brosaplanella** [reply]
It's hard to tell where the problem is just with the log. Can you share a minimum working example?

SEI growth and lithium plating can be enabled through model options, you can find an example here: https://docs.pybamm.org/en/latest/source/examples/notebooks/models/coupled-degradation.html
